plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,126 @@
+        extends AbstractTestQueryFramework
+{
+    protected ConnectionCountingConnectionFactory connectionFactory;
+
+    @BeforeClass
+    public void verifySetup()
+    {
+        // Test expects connectionFactory to be provided with AbstractTestQueryFramework.createQueryRunner implementation
+        requireNonNull(connectionFactory, "connectionFactory is null");
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected void assertJdbcConnections(@Language("SQL") String query, int expectedJdbcConnectionsCount, Optional<String> errorMessage)
+    {
+        int before = connectionFactory.openConnections.get();
+        if (errorMessage.isPresent()) {
+            assertQueryFails(query, errorMessage.get());
+        }
+        else {
+            getQueryRunner().execute(query);
+        }
+        int after = connectionFactory.openConnections.get();
+        assertThat(after - before).isEqualTo(expectedJdbcConnectionsCount);
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    protected static class ConnectionCountingConnectionFactory
+            implements ConnectionFactory
+    {
+        // Map from connection to a fake exception (holds stacktrace) pointing to the place where the connection was created
+        private final Map<Connection, Exception> connectionCreations = synchronizedMap(new IdentityHashMap<>());
+        private final AtomicInteger openConnections = new AtomicInteger();
+        private final ConnectionFactory delegate;
+
+        public ConnectionCountingConnectionFactory(DriverConnectionFactory delegate)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        public Connection openConnection(ConnectorSession session)
+                throws SQLException
+        {
+            openConnections.incrementAndGet();
+            Connection connection = delegate.openConnection(session);
+            Exception previous = connectionCreations.put(connection, new Exception("STACKTRACE"));
+            if (previous != null) {
+                IllegalStateException exception = new IllegalStateException("Same connection %s returned twice from delegate".formatted(connection));
+                exception.addSuppressed(previous);
+                throw exception;
+            }
+            return new ForwardingConnection()
+            {
+                private volatile boolean closed;
+
+                @Override
+                protected Connection delegate()
+                {
+                    return connection;
+                }
+
+                @Override
+                public void close()
+                        throws SQLException
+                {
+                    if (closed) {
+                        return;
+                    }
+                    closed = true;
+                    connectionCreations.remove(connection);
+                    super.close();
```

Comment:
To make sure that `closed` and `connectionCreations` are in a good shape. However it does not matter much. Tests will fail in case of any exception here.

Reply:
I think it does not matter too.